### PR TITLE
Prometheus Metrics

### LINKED ISSUE #143: Prometheus metrics
### Environment
* Nautobot version:  1.5.13 (to get `metrics.py` support)
* nautobot-plugin-device-lifecycle-mgmt version:  latest

### Proposed Functionality

Expose Prometheus metrics for this app. Examples:

#### Device Software Validation Report

This GUI view shows a graph that details how many devices have valid platforms. Metrics that _could_ be generated from this:

- `nautobot_device_lifecycle_compliant_by_platform_total{platform="Arista EOS", status="compliant"} 160.0`
- `nautobot_device_lifecycle_compliant_by_device_type_total{device_type="Catalyst 6509-E", status="compliant"} 35.0`

Further metrics could be useful, these are just examples I was able to think of.

### Use Case

Allow users of the platform to generate time series from their Nautobot instance. In the end this could for example lead to creation of NOC dashboards or alerting based on this data.

### PR DESCRIPTION
**Pull request by alhogan**
Closes #143
@zackt25 and I added some basic metrics for Lifecycle Management.

```shell
# HELP nautobot_lcm_compliance_by_device_type_total Number of devices that have valid/invalid software by device_type
# TYPE nautobot_lcm_compliance_by_device_type_total gauge
nautobot_lcm_compliance_by_device_type_total{device_type="ASR-9903",is_valid="True"} 3.0
nautobot_lcm_compliance_by_device_type_total{device_type="ASR-9903",is_valid="False"} 1.0
nautobot_lcm_compliance_by_device_type_total{device_type="IntelServer",is_valid="True"} 2.0
nautobot_lcm_compliance_by_device_type_total{device_type="IntelServer",is_valid="False"} 2.0
nautobot_lcm_compliance_by_device_type_total{device_type="WS-C3560CX-12PC-S",is_valid="True"} 2.0
nautobot_lcm_compliance_by_device_type_total{device_type="WS-C3560CX-12PC-S",is_valid="False"} 2.0
nautobot_lcm_compliance_by_device_type_total{device_type="mx-960",is_valid="True"} 3.0
nautobot_lcm_compliance_by_device_type_total{device_type="mx-960",is_valid="False"} 1.0
# HELP nautobot_lcm_valid_by_inventory_item_total Number of devices that have valid/invalid software by inventory item
# TYPE nautobot_lcm_valid_by_inventory_item_total gauge
nautobot_lcm_valid_by_inventory_item_total{inventory_item="Cisco-FEX-port1",is_valid="True"} 0.0
nautobot_lcm_valid_by_inventory_item_total{inventory_item="Cisco-FEX-port1",is_valid="False"} 1.0
```

### CODE REVIEW

**Comment by progala** [comment]
@bradh11 Sure, I'll have a look today.

**Comment by jvanderaa** [comment]
We should look to take nautobot out all together on pyproject.toml if doing anything to it. 


Is there something in particular that we need to be at 1.5.13??

Let's also remove the changes to the migration files.

**Comment by alhogan** [comment]
I suspect adding `nautobot-capacity-metrics` as a dependency is what caused the flake8 downgrade. I'm happy to remove the changes outside of `metrics.py` but my understanding is that the metrics will not be available without nautobot >= 1.5.13 and the capacity metrics plugin.

**Review by alhogan — COMMENTED** [review]
All changes outside of `metrics.py` have been reverted

File: nautobot_device_lifecycle_mgmt/metrics.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+        "nautobot_lcm_valid_by_device_type_total",
+        "Number of devices that have valid software by device_type",
+        labels=["device_type"],
+    )
+
+    invalid_device_software_gauge = GaugeMetricFamily(
+        "nautobot_lcm_invalid_by_device_type_total",
+        "Number of devices that have invalid software by device_type",
+        labels=["device_type"],
+    )
```
Comment (Kircheneer) [inline nautobot_device_lifecycle_mgmt/metrics.py:29]: My vote would be to put both of these into one metric and have a valid/invalid label instead.

File: nautobot_device_lifecycle_mgmt/metrics.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+            ),
+        )
+        invalid_device_software_gauge.add_metric(
+            labels=[model],
+            value=(
+                DeviceSoftwareValidationResult.objects.filter(
+                    device__device_type__model=model, is_validated=False
+                ).count()
+            ),
+        )
```
Comment (Kircheneer) [inline nautobot_device_lifecycle_mgmt/metrics.py:48]: > if not no gauge will be created

According to [best practices](https://prometheus.io/docs/practices/instrumentation/#avoid-missing-metrics) you should _always_ provide a metric, even if the sample is just set to 0 and no labels are present.

File: nautobot_device_lifecycle_mgmt/metrics.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+        "nautobot_lcm_valid_by_inventory_item_total",
+        "Number of devices that have valid software by inventory item",
+        labels=["inventory_item"],
+    )
+
+    invalid_item_software_gauge = GaugeMetricFamily(
+        "nautobot_lcm_invalid_by_inventory_item_total",
+        "Number of devices that have invalid software by inventory item",
+        labels=["inventory_item"],
+    )
```
Comment (Kircheneer) [inline nautobot_device_lifecycle_mgmt/metrics.py:77]: I would vote to also condense these into one metric through an additional label

File: nautobot_device_lifecycle_mgmt/metrics.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+            ),
+        )
+        invalid_item_software_gauge.add_metric(
+            labels=[model],
+            value=(
+                InventoryItemSoftwareValidationResult.objects.filter(
+                    inventory_item__part_id=model, is_validated=False
+                ).count()
+            ),
+        )
```
Comment (Kircheneer) [inline nautobot_device_lifecycle_mgmt/metrics.py:96]: Also needs a case for `inventory_item_validation = []`

File: nautobot_device_lifecycle_mgmt/metrics.py [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+        "Number of inventory items that are in lifecycle management report",
+        labels=["inventory_items"],
+    )
+
+    total_inventory_item_software_gauge.add_metric(
+        labels=["Total LCM Validated Inventory Item Count"],
+        value=(InventoryItemSoftwareValidationResult.objects.all().count()),
+    )
+
+    yield total_inventory_item_software_gauge
```
Comment (Kircheneer) [inline nautobot_device_lifecycle_mgmt/metrics.py:133]: Counting of model instances should be available in core directly whenever I get [this PR](https://github.com/nautobot/nautobot-plugin-capacity-metrics/pull/46) merged into core - so this metric is not necessary.

**Review by Kircheneer — CHANGES_REQUESTED** [review]

**Comment by progala** [comment]
@alhogan @Kircheneer My understanding is that if you're running Nautobot < 1.5.13 you simply won't have access to the metrics but the plugins should still work. If that's the case we should keep the minimum requirement at 1.4.0 and make sure docs and release notes explain what's required to have metrics.

**Comment by progala** [comment]
I'm working on PR which will use querysets instead of loops. One question I have is around the treatment of inventory items that have no part_id defined, i.e. where the value of part_id is an empty string. Should we exclude those completely from the queries? @alhogan @Kircheneer @nkallergis

**Comment by nkallergis** [comment]
That would be my approach since (if I'm not mistaken) there's no means to assign an LCM notice on them. That said, keep in mind that the other PR (#151, now merged) did include them.

**Comment by progala** [comment]
@alhogan @nkallergis @Kircheneer I've converted the logic to use the query annotations. I'd be grateful if you could review the PR as it is now and we'll hopefully be able to merge it soon.

**Review by nkallergis — APPROVED** [review]
That's actually wonderful and so easy to read! Makes me think that a) I need to learn better Django and b) maybe we need to revisit the metrics implementation of other plugins to align with this approach 🤔

**Review by Kircheneer — APPROVED** [review]
LGTM

**Pull request merged**